core/src/main/java/org/apache/calcite/rel/rules/CoreRules.java

```diff
@@ -288,8 +290,16 @@ private CoreRules() {}
   /** Rule that pushes a {@link LogicalFilter}
    * past a {@link LogicalTableFunctionScan}. */
   public static final FilterTableFunctionTransposeRule
-      FILTER_TABLE_FUNCTION_TRANSPOSE =
-      FilterTableFunctionTransposeRule.Config.DEFAULT.toRule();
+      LOGICAL_FILTER_LOGICAL_TABLE_FUNCTION_TRANSPOSE = Config.DEFAULT
+          .withOperandFor(LogicalFilter.class, LogicalTableFunctionScan.class)
```

Comment:
Maybe I am missing something here, but both `LogicalFilter` and `LogicalTableFunctionScan` are subclasses of `Filter` and `TableFunctionScan` (respectively), why do we need to duplicate the rule matching condition here? I don't recall us doing it for other Calcite rules which most probably have the same problem (I recall a few changes lifting the Logical operator to the parent class some time ago which didn't require any introduction of new methods nor deprecating anything, but I can't find it at the moment).

If we provide the rule operand matching for `X` won't it work and get applied to `LogicalX`? In that case you don't need to introduce the `LOGICAL_` elements.

Reply:
Because I saw Julian mention this statement in JIRA: "Perhaps retain an instance of the rule that continues to match only LogicalFilter and LogicalTableScan, and also add a new instance that is generic." Therefore, I kept a Logical_X rule.

Reply:
Thanks for the explanation @xiedeyantu, maybe @julianhyde can elaborate more on this? I still don't get the rationale behind the request, I don't feel strongly about it, it's more to understand better.